File: src/main/java/me/ag2s/epublib/domain/LazyResource.java
package me.ag2s.epublib.domain;

import java.io.ByteArrayInputStream;
import java.io.IOException;
import java.io.InputStream;
import me.ag2s.epublib.util.IOUtil;

public class LazyResource extends Resource {
   private static final long serialVersionUID = 5089400472352002866L;
   private final String TAG;
   private final LazyResourceProvider resourceProvider;
   private final long cachedSize;

   public LazyResource(LazyResourceProvider resourceProvider, String href) {
      this(resourceProvider, -1L, href);
   }

   public LazyResource(LazyResourceProvider resourceProvider, String href, String originalHref) {
      this(resourceProvider, -1L, href, originalHref);
   }

   public LazyResource(LazyResourceProvider resourceProvider, long size, String href) {
      super((String)null, (byte[])null, href, MediaTypes.determineMediaType(href));
      this.TAG = this.getClass().getName();
      this.resourceProvider = resourceProvider;
      this.cachedSize = size;
   }

   public LazyResource(LazyResourceProvider resourceProvider, long size, String href, String originalHref) {
      super((String)null, (byte[])null, href, (String)originalHref, (MediaType)MediaTypes.determineMediaType(href));
      this.TAG = this.getClass().getName();
      this.resourceProvider = resourceProvider;
      this.cachedSize = size;
   }

   public InputStream getInputStream() throws IOException {
      return (InputStream)(this.isInitialized() ? new ByteArrayInputStream(this.getData()) : this.resourceProvider.getResourceStream(this.originalHref));
   }

   public void initialize() throws IOException {
      this.getData();
   }

   public byte[] getData() throws IOException {
      if (this.data == null) {
         InputStream in = this.resourceProvider.getResourceStream(this.originalHref);
         byte[] readData = IOUtil.toByteArray(in, (int)this.cachedSize);
         if (readData == null) {
            throw new IOException("Could not load the contents of resource: " + this.getHref());
         }

         this.data = readData;
         in.close();
      }

      return this.data;
   }

   public void close() {
      if (this.resourceProvider != null) {
         this.data = null;
      }

   }

   public boolean isInitialized() {
      return this.data != null;
   }

   public long getSize() {
      return this.data != null ? (long)this.data.length : this.cachedSize;
   }
}

File: src/main/kotlin/me/ag2s/epublib/domain/LazyResource.kt
package me.ag2s.epublib.domain

import java.io.ByteArrayInputStream
import java.io.IOException
import java.io.InputStream
import me.ag2s.epublib.util.IOUtil

class LazyResource : Resource {
    private val tag = javaClass.name
    private val resourceProvider: LazyResourceProvider
    private val cachedSize: Long
    constructor(resourceProvider: LazyResourceProvider, href: String?) : this(resourceProvider, -1L, href)
    constructor(resourceProvider: LazyResourceProvider, href: String?, originalHref: String?) : this(resourceProvider, -1L, href, originalHref)
    constructor(resourceProvider: LazyResourceProvider, size: Long, href: String?) : super(null, null, href, MediaTypes.determineMediaType(href)) { this.resourceProvider = resourceProvider; cachedSize = size }
    constructor(resourceProvider: LazyResourceProvider, size: Long, href: String?, originalHref: String?) : super(null, null, href, originalHref, MediaTypes.determineMediaType(href)) { this.resourceProvider = resourceProvider; cachedSize = size }
    @Throws(IOException::class)
    override fun getInputStream(): InputStream = if (isInitialized()) ByteArrayInputStream(getData()) else resourceProvider.getResourceStream(originalHref)
    @Throws(IOException::class)
    fun initialize() { getData() }
    @Throws(IOException::class)
    override fun getData(): ByteArray {
        if (data == null) {
            val input = resourceProvider.getResourceStream(originalHref)
            val readData = IOUtil.toByteArray(input, cachedSize.toInt())
            if (readData == null) throw IOException("Could not load the contents of resource: ${getHref()}")
            data = readData
            input.close()
        }
        return data!!
    }
    override fun close() { data = null }
    fun isInitialized(): Boolean = data != null
    override fun getSize(): Long = data?.size?.toLong() ?: cachedSize
    companion object { private const val serialVersionUID = 5089400472352002866L }
}
